crates/jit/src/compiler.rs

```diff
@@ -120,66 +107,63 @@ impl Compiler {
-        ) = match self.strategy {
-            // For now, interpret `Auto` as `Cranelift` since that's the most stable
-            // implementation.
-            CompilationStrategy::Auto | CompilationStrategy::Cranelift => {
-                wasmtime_environ::cranelift::Cranelift::compile_module(translation, &*self.isa)
-            }
-            #[cfg(feature = "lightbeam")]
-            CompilationStrategy::Lightbeam => {
-                wasmtime_environ::lightbeam::Lightbeam::compile_module(translation, &*self.isa)
+        cfg_if::cfg_if! {
+            if #[cfg(feature = "parallel-compilation")] {
+                use rayon::prelude::*;
+                let iter = translation.function_body_inputs
+                    .iter()
+                    .collect::<Vec<_>>()
+                    .into_par_iter();
+            } else {
+                let iter = translation.function_body_inputs.iter();
             }
         }
-        .map_err(SetupError::Compile)?;
-
-        let dwarf_sections = if translation.debuginfo.is_some() && !compilation.is_empty() {
-            let unwind_info = compilation.unwind_info();
+        let mut funcs = iter
+            .map(|(index, func)| {
+                self.compiler
+                    .compile_function(translation, index, func, &*self.isa)
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .collect::<CompiledFunctions>();
+
+        let dwarf_sections = if translation.debuginfo.is_some() && !funcs.is_empty() {
             transform_dwarf_data(
                 &*self.isa,
                 &translation.module,
                 translation.debuginfo.as_ref().unwrap(),
-                &address_transform,
-                &value_ranges,
-                stack_slots,
-                unwind_info,
+                &funcs,
             )?
         } else {
             vec![]
         };
 
-        let (obj, unwind_info) = build_object(
-            &*self.isa,
-            &translation.module,
-            compilation,
-            relocations,
-            dwarf_sections,
-        )?;
+        let (obj, unwind_info) =
+            build_object(&*self.isa, &translation.module, &funcs, dwarf_sections)?;
+
+        // Clear out fields which are no longer needed for each function,
+        // releasing their memory back to the system. We may want to prefer to
+        // do this with `Option` to prevent access in the future as well. These
+        // fields were all used to construct the object file and/or debug info,
+        // but from this point on they're no longer needed.
+        for func in funcs.values_mut() {
+            func.body = Default::default();
+            func.jt_offsets = Default::default();
+            func.relocations = Default::default();
+            func.value_labels_ranges = Default::default();
+            func.stack_slots = Default::default();
+            func.unwind_info = Default::default();
+        }
```

Comment:
Is it possible to add "CompiledFunction"-lite-like structure to use it with `Compilation`? It does not look like the code above will be easy to maintain.

Reply:
Yeah that's a good point, I went ahead and updated `Compilation` to have specific accessors for what's needed by wasmtime and now it no longer stores all these fields that are emptied out here.